src/drivers/magnetometer/memsic/mmc5983ma/mmc5983ma.cpp

```diff
@@ -0,0 +1,215 @@
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+#include "mmc5983ma.h"
+
+using namespace time_literals;
+
+MMC5983MA::MMC5983MA(device::Device *interface, const I2CSPIDriverConfig &config) :
+	I2CSPIDriver(config),
+	_interface(interface),
+	_px4_mag(interface->get_device_id(), config.rotation),
+	_sample_perf(perf_alloc(PC_ELAPSED, "mmc5983ma_read")),
+	_comms_errors(perf_alloc(PC_COUNT, "mmc5983ma_comms_errors"))
+{}
+
+MMC5983MA::~MMC5983MA()
+{
+	perf_free(_sample_perf);
+	perf_free(_comms_errors);
+	delete _interface;
+}
+
+int MMC5983MA::init()
+{
+	// Start with a reset of the chip
+	write_register(MMC5983MA_ADDR_CTRL_REG2, MMC5983MA_CTRL_REG1_SW_RESET);
+	px4_usleep(20_ms);
+
+	// Triggered mode with degaussing (averaging two samples), 18-bit resolution, automatic set/reset
+	write_register(MMC5983MA_ADDR_CTRL_REG0, MMC5983MA_CTRL_REG0_AUTO_SR_EN);
+	write_register(MMC5983MA_ADDR_CTRL_REG1, MMC5983MA_CTRL_REG1_BW_200HZ);
+	write_register(MMC5983MA_ADDR_CTRL_REG2, MMC5983MA_CTRL_REG2_EN_PRD_SET);
+
+	ScheduleNow();
+
+	return PX4_OK;
+}
+
+void MMC5983MA::RunImpl()
+{
+	switch (_state) {
+	case State::Measure: {
+			auto start_time = hrt_absolute_time();
+
+			write_register(MMC5983MA_ADDR_CTRL_REG0, MMC5983MA_CTRL_REG0_TM_M);
+
+			_collect_retries = 0;
+			_state = State::Collect;
+
+			auto elapsed = hrt_elapsed_time(&start_time);
+
+			if (elapsed > 5_ms) {
+				ScheduleNow();
+
+			} else {
+				ScheduleDelayed(5_ms - elapsed);
+			}
+
+			return;
+		}
+
+	case State::Collect: {
+			auto start_time = hrt_absolute_time();
+
+			uint8_t status = read_register(MMC5983MA_ADDR_STATUS_REG);
+
+			if (status & MMC5983MA_STATUS_REG_MEAS_M_DONE) {
+				SensorData data = {};
+
+				if (read_register_block(&data) != PX4_OK) {
+					PX4_DEBUG("read failed");
+					perf_count(_comms_errors);
+					_state = State::Measure;
+					_sample_index = 0;
+					ScheduleDelayed(100_ms);
+					return;
+				}
+
+				// Measurement available
+				_measurements[_sample_index] = data;
```

Comment:
Why are you taking 2 measurements before publishing?

Reply:
See page 18 of the datasheet
**USING SET AND RESET TO REMOVE BRIDGE OFFSET**
https://www.memsic.com/Public/Uploads/uploadfile/files/20220119/MMC5983MADatasheetRevA.pdf